core/arch/arm/kernel/secure_partition.c

```diff
@@ -526,6 +570,155 @@ static TEE_Result handle_fdt_dev_regions(struct sp_ctx *ctx, void *fdt)
+			alloc_needed = true;
+
+		/* Size of memory region as count of 4K pages */
+		if (sp_dt_get_u32(fdt, subnode, "pages-count", &pages_cnt)) {
+			EMSG("Mandatory field is missing: pages-count");
+			return TEE_ERROR_BAD_FORMAT;
+		}
+
+		if (MUL_OVERFLOW(pages_cnt, SMALL_PAGE_SIZE, &size))
+			return TEE_ERROR_OVERFLOW;
+
+		/*
+		 * Memory region attributes:
+		 * - Instruction/data access permissions
+		 * - Cacheability/shareability attributes
+		 * - Security attributes
+		 *
+		 * Cacheability/shareability attributes can be ignored for now.
+		 * OP-TEE only supports a single type for normal cached memory
+		 * and currently there is no use case that would require to
+		 * change this.
+		 */
+		if (sp_dt_get_u32(fdt, subnode, "attributes", &attributes)) {
+			EMSG("Mandatory field is missing: attributes");
+			return TEE_ERROR_BAD_FORMAT;
+		}
+
+		/* Check instruction and data access permissions */
+		switch (attributes & SP_MANIFEST_ATTR_RWX) {
+		case SP_MANIFEST_ATTR_RO:
+			perm = TEE_MATTR_UR;
+			break;
+		case SP_MANIFEST_ATTR_RW:
+			perm = TEE_MATTR_URW;
+			break;
+		case SP_MANIFEST_ATTR_RX:
+			perm = TEE_MATTR_URX;
+			break;
+		default:
+			EMSG("Invalid memory access permissions");
+			return TEE_ERROR_BAD_FORMAT;
+		}
+
+		/*
+		 * The SP is a secure endpoint, security attribute can be
+		 * secure or non-secure.
+		 * The SPMC cannot allocate non-secure memory, i.e. if the base
+		 * address is missing this attribute must be secure.
+		 */
+		if (attributes & SP_MANIFEST_ATTR_NSEC) {
+			if (alloc_needed) {
+				EMSG("Invalid memory security attribute");
+				return TEE_ERROR_BAD_FORMAT;
+			}
+			is_secure = false;
+		}
+
+		if (alloc_needed) {
+			/* Base address is missing, we have to allocate */
+			mm = tee_mm_alloc(&tee_mm_sec_ddr, size);
+			if (!mm)
+				return TEE_ERROR_OUT_OF_MEMORY;
+
+			base_addr = tee_mm_get_smem(mm);
+		}
+
+		m = sp_mem_new_mobj(pages_cnt, TEE_MATTR_MEM_TYPE_CACHED,
+				    is_secure);
+		if (!m) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			goto err_mm_free;
+		}
+
+		res = sp_mem_add_pages(m, &idx, base_addr, pages_cnt);
+		if (res) {
+			mobj_put(m);
+			goto err_mm_free;
+		}
+
+		res = vm_map(&ctx->uctx, &va, size, perm, 0, m, 0);
+		mobj_put(m);
+		if (res)
+			goto err_mm_free;
+
+		/*
+		 * Overwrite the memory region's base address in the fdt with
+		 * the VA. This fdt will be passed to the SP.
+		 * If the base-address field was not present in the original
+		 * fdt, this function will create it. This doesn't cause issues
+		 * since the necessary extra space has been allocated when
+		 * opening the fdt.
+		 */
+		res = fdt_setprop_u64(fdt, subnode, "base-address", va);
+
+		/*
+		 * Unmap the region if the overwrite failed since the SP won't
+		 * be able to access it without knowing the VA.
+		 */
+		if (res) {
+			vm_unmap(&ctx->uctx, va, size);
+			goto err_mm_free;
+		}
+	}
+
+	return TEE_SUCCESS;
+
+err_mm_free:
+	/* If mm is NULL, tee_mm_free() will just return */
```

Comment:
i think you can discard this inline comment.

Reply:
Done